### ISSUE TITLE
Training speed on M4 is very slow

### ISSUE THREAD

**Issue by Avencheng** [ultralytics/ultralytics#22778]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I used to train YOLO on an Ubuntu environment with Nvidia RTX A2000. Recently, I switched to using a Apple M4 Max as my training machine, but I found that the training speed on the M4 is very slow—possibly only about 1/6 of the speed of an Nvidia GPU.

Below is my training environment：
```
conda create -n MPS-312 python=3.12
conda activate MPS-312 
pip3 install torch torchvision
pip install ultralytics==8.3.231
```

My training parameters
```
from ultralytics import YOLO
from ultralytics.utils.autobatch import autobatch
import torch

torch.mps.set_per_process_memory_fraction(0.8)
torch.mps.empty_cache()

# Load a model
model = YOLO('yolo11s-cls.pt') 

# Train the model
batch_size = autobatch(model=model, imgsz=imgsz, fraction=0.6) 
results = model.train(data="EBCNews_images1", epochs=450, imgsz=512, device="mps", cache=True, batch=batch_size, workers=4, fraction=0.7, half=False, cos_lr=True, hsv_h=0.0, hsv_s=0.0, hsv_v=0.0, fliplr=0, mosaic=0, erasing=0)
```

A little bit of my training process, the iterations per second is only 1.8it/s ...
```
Epoch    GPU_mem       loss  Instances       Size
1/450       3.5G      1.681         12        512: 100% ━━━━━━━━━━━━ 148/148 1.8it/s 1:23
         classes   top1_acc   top5_acc: 100% ━━━━━━━━━━━━ 30/30 1.2it/s 25.5s
             all      0.661      0.764

Epoch    GPU_mem       loss  Instances       Size
2/450      3.37G     0.5229         12        512: 100% ━━━━━━━━━━━━ 148/148 1.8it/s 1:23
         classes   top1_acc   top5_acc: 100% ━━━━━━━━━━━━ 30/30 1.2it/s 25.4s
             all      0.681      0.772

Epoch    GPU_mem       loss  Instances       Size
3/450      3.37G     0.4341         12        512: 100% ━━━━━━━━━━━━ 148/148 1.8it/s 1:23
         classes   top1_acc   top5_acc: 100% ━━━━━━━━━━━━ 30/30 1.2it/s 25.4s
             all      0.689      0.771
```

I'm not sure if this speed is normal. Is there any way to increase the training speed? Thanks!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Avencheng, thanks for opening this issue and for sharing your detailed setup and logs 🚀  
This is an automated response to help move things forward quickly — an Ultralytics engineer will also review your issue and follow up soon.

From your description, this looks like a custom training performance ❓ question on Apple M‑series (MPS), not a straightforward bug report. To help the team better understand and reproduce the behavior on Apple silicon, please provide a bit more information:
- Exact YOLO model and task you are training (you mentioned `yolo11s-cls.pt`, please confirm if this is classification only and whether any custom modifications were made)
- Dataset details: number of images, number of classes, and a rough idea of image resolution before/after preprocessing
- Your exact `imgsz` value (since it’s used in `autobatch` but not shown in your snippet)
- Whether you see similar speeds when:
  - Reducing `imgsz`
  - Reducing `workers` to `0` or `1`
  - Disabling `cache=True`
- Full training command or script (including how you launch Python/conda) and full console output for at least a few epochs

In parallel, we strongly recommend checking the Docs and environment first, as many performance‑related questions are covered there and often resolved by upgrading or adjusting configuration ⚙️

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Before the team investigates further, please upgrade and verify the issue still occurs on the latest release, as many performance and MPS-related improvements land over time.

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a **Python>=3.8** environment with **PyTorch>=1.8** to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

After upgrading, please:
1. Re-run your training with the same configuration on your M4 Max
2. Share:
   - The first 3–5 epochs of logs
   - Reported `it/s`
   - Any warnings printed related to MPS or dtype/precision

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

If you have access to one of these GPU environments, a quick comparison run (same model, same dataset, similar `imgsz` and batch size) plus logs from both M4 and GPU will be very helpful to the engineers in assessing whether your M4 performance is within expectations or indicates an issue.

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Once you share the extra details and confirm behavior on the latest version, it will greatly help the team investigate your M4 Max training speed 🧠💻

**Comment by Y-T-G** [comment]
What's the batch size?

**Comment by Avencheng** [comment]
The batch size is 16

**Comment by glenn-jocher** [comment]
On Apple M‑series, MPS training is generally quite a bit slower than CUDA, so 1.8 it/s with `batch=16` at `imgsz=512` is unfortunately in a normal range; to squeeze more speed out you can try manually increasing `batch` (since `autobatch` only optimizes CUDA), enabling mixed precision with `amp=True`, and raising `workers`, as described in the [Apple Silicon MPS training guide](https://docs.ultralytics.com/modes/train/#apple-silicon-mps-training). If it still feels abnormally slow after that, please run `yolo checks` (or `!yolo checks` in a notebook) and share the output plus your dataset size/classes so we can compare against a known‑good setup.

**Comment by nailson-landim** [comment]
Correct me If i'm wrong @glenn-jocher but that requires quite a lot of optimizations, like writing Metal/MLX code to maybe improve performance.

**Comment by glenn-jocher** [comment]
You’re basically right that large speedups would need work in the underlying MPS/Metal stack (i.e., PyTorch/Apple), not just Ultralytics—our side just uses the standard `device="mps"` path—so beyond model/dataloader choices (e.g., smaller `imgsz`, lighter `yolo11n`/`yolo11s`, subset training via `fraction<1`, etc. as outlined in the [training tips guide](https://docs.ultralytics.com/guides/model-training-tips/)), there isn’t a repo-level switch that will make MPS match CUDA.

**Comment by Avencheng** [comment]
@glenn-jocher Thank you for your reply. I tried the settings you suggested, but the speed didn’t improve noticeably. For now, I’ve decided not to train models on the Mac Studio and will switch back to training with CUDA instead, this would be easier for me.

**Comment by glenn-jocher** [comment]
That makes sense—at the moment PyTorch `mps` training often can’t match CUDA throughput, so switching back to your RTX for training is usually the most efficient path; your M4 Max can still be a great local box for `predict()`/deployment prep via [Export mode](https://docs.ultralytics.com/modes/export/), and if you ever want us to dig into MPS performance again please open a new issue with `yolo checks` output plus the full training log so we can rule out any environment/dataloader bottlenecks.

**Comment by Sant16** [comment]
What computer do the technology leaders at Ultralytics currently use in their day-to-day work?

**Comment by nailson-landim** [comment]
Its at their employee handbook somewhere. Standard apple m5 machines.

They train on nVidia GPUs.


Em qua., 18 de mar. de 2026 às 23:10, Sant ***@***.***>
escreveu:

**Comment by glenn-jocher** [comment]
It varies by role rather than one fixed machine, but broadly day-to-day development is often on Apple Silicon while training and benchmarking are typically done on NVIDIA CUDA systems; since that’s outside this closed MPS issue, please continue general questions in [Ultralytics Discussions](https://github.com/orgs/ultralytics/discussions).

**Comment by rolyantrauts** [comment]
Dunno why some are ignoring the suggestion to increase the batchsize as its already been said (since autobatch only optimizes CUDA) pulling only 16 samples each step batchsize=16 is very small reducing parallelism and likely 512 with the base is prob max.
Just check your ram usage as that is the trade off of batch size as autobatch maximises batchsize to the available vram of CUDA.
You will get somewhere between 3050/3060 speed with a m4 mini base version

**Comment by glenn-jocher** [comment]
Yes, that’s the key detail here: [`autobatch()` is intended for CUDA and falls back to the default batch on `cpu`/`mps`](https://docs.ultralytics.com/reference/utils/autobatch/), so on Apple Silicon the practical approach is a manual batch sweep like `16 -> 32 -> 64` while watching unified memory and swap, since once memory pressure starts, throughput usually drops instead of improving.
